File: src-tauri/src/update_checker.rs
// src-tauri/src/update_checker.rs
//
// Equivalent to electron/updateChecker.ts

use anyhow::{Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};

const GITHUB_OWNER: &str = "Kim2091";
const GITHUB_REPO: &str = "vapourkit";

#[derive(Debug, Deserialize)]
struct GitHubRelease {
    tag_name: String,
    name: String,
    body: String,
    html_url: String,
    published_at: String,
    prerelease: bool,
    draft: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct UpdateInfo {
    pub available: bool,
    pub current_version: String,
    pub latest_version: String,
    pub release_url: String,
    pub changelog: String,
    pub published_at: String,
}

pub fn get_releases_page_url() -> String {
    format!("https://github.com/{}/{}/releases", GITHUB_OWNER, GITHUB_REPO)
}

pub fn get_app_version() -> String {
    env!("CARGO_PKG_VERSION").to_string()
}

/// Compare two semver strings. Returns true if latest > current.
fn is_newer(current: &str, latest: &str) -> bool {
    compare_versions(latest, current) > 0
}

fn compare_versions(v1: &str, v2: &str) -> i32 {
    let clean1 = v1.trim_start_matches('v');
    let clean2 = v2.trim_start_matches('v');

    let parts1: Vec<u64> = clean1
        .split('.')
        .map(|p| p.parse().unwrap_or(0))
        .collect();
    let parts2: Vec<u64> = clean2
        .split('.')
        .map(|p| p.parse().unwrap_or(0))
        .collect();

    let len = parts1.len().max(parts2.len());
    for i in 0..len {
        let a = parts1.get(i).copied().unwrap_or(0);
        let b = parts2.get(i).copied().unwrap_or(0);
        if a > b {
            return 1;
        }
        if a < b {
            return -1;
        }
    }
    0
}

pub async fn check_for_updates() -> Result<UpdateInfo> {
    let current_version = get_app_version();
    log::info!("Checking for updates, current version: {}", current_version);

    let url = format!(
        "https://api.github.com/repos/{}/{}/releases/latest",
        GITHUB_OWNER, GITHUB_REPO
    );

    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .user_agent("vapourkit-app")
        .build()?;

    let release: GitHubRelease = client
        .get(&url)
        .header("Accept", "application/vnd.github.v3+json")
        .send()
        .await
        .context("GitHub API request")?
        .json()
        .await
        .context("parse GitHub response")?;

    if release.draft || release.prerelease {
        log::info!("Latest release is draft/prerelease, skipping");
        return Ok(UpdateInfo {
            available: false,
            current_version: current_version.clone(),
            latest_version: current_version,
            release_url: String::new(),
            changelog: String::new(),
            published_at: String::new(),
        });
    }

    let latest = release.tag_name.trim_start_matches('v').to_string();
    let available = is_newer(&current_version, &latest);

    log::info!(
        "Latest version: {}, available: {}",
        latest,
        available
    );

    Ok(UpdateInfo {
        available,
        current_version,
        latest_version: latest,
        release_url: release.html_url,
        changelog: release.body,
        published_at: release.published_at,
    })
}

File: electron/updateChecker.ts
import axios from 'axios';
import { app } from 'electron';
import { logger } from './logger';

interface GitHubRelease {
  tag_name: string;
  name: string;
  body: string;
  html_url: string;
  published_at: string;
  prerelease: boolean;
  draft: boolean;
}

export interface UpdateInfo {
  available: boolean;
  currentVersion: string;
  latestVersion: string;
  releaseUrl: string;
  changelog: string;
  publishedAt: string;
}

const GITHUB_OWNER = 'Kim2091';
const GITHUB_REPO = 'vapourkit';
const GITHUB_API_URL = `https://api.github.com/repos/${GITHUB_OWNER}/${GITHUB_REPO}/releases/latest`;

export interface ParsedVersion {
  core: number[];
  prerelease: string | null;
}

/**
 * Parses "v0.16.1-nightly.2026-05-13" into { core: [0, 16, 1], prerelease: 'nightly.2026-05-13' }
 */
export function parseVersion(version: string): ParsedVersion {
  const clean = version.replace(/^v/, '');
  const dashIndex = clean.indexOf('-');
  const corePart = dashIndex === -1 ? clean : clean.slice(0, dashIndex);
  const prerelease = dashIndex === -1 ? null : clean.slice(dashIndex + 1);

  const core = corePart.split('.').map(part => {
    const num = parseInt(part, 10);
    return Number.isNaN(num) ? 0 : num;
  });

  return { core, prerelease };
}

/**
 * Returns true when the latest (stable) release is an actual upgrade from the
 * current version.
 *
 * Only the numeric version core is compared. Nightly builds carry a prerelease
 * suffix (e.g. 0.16.1-nightly.2026-05-13) that the old comparison mangled to
 * NaN→0, making the same-version stable release look newer — nightly users were
 * endlessly prompted to "update" to the stable build they were cut from. A
 * nightly is only offered an update once a stable release with a strictly newer
 * base version exists.
 */
export function isUpdateAvailable(currentVersion: string, latestVersion: string): boolean {
  const current = parseVersion(currentVersion);
  const latest = parseVersion(latestVersion);

  for (let i = 0; i < Math.max(current.core.length, latest.core.length); i++) {
    const currentNum = current.core[i] ?? 0;
    const latestNum = latest.core[i] ?? 0;

    if (latestNum > currentNum) return true;
    if (latestNum < currentNum) return false;
  }

  return false;
}

/**
 * Checks for updates by comparing current version with latest GitHub release
 */
export async function checkForUpdates(): Promise<UpdateInfo> {
  try {
    logger.info('Checking for updates...');
    
    // Get current version from package.json
    const currentVersion = app.getVersion();
    logger.info(`Current version: ${currentVersion}`);
    
    // Fetch latest release from GitHub
    const response = await axios.get<GitHubRelease>(GITHUB_API_URL, {
      headers: {
        'Accept': 'application/vnd.github.v3+json',
        'User-Agent': 'vapourkit-app'
      },
      timeout: 10000 // 10 second timeout
    });
    
    const release = response.data;
    
    // Skip draft and prerelease versions
    if (release.draft || release.prerelease) {
      logger.info('Latest release is draft or prerelease, skipping');
      return {
        available: false,
        currentVersion,
        latestVersion: currentVersion,
        releaseUrl: '',
        changelog: '',
        publishedAt: ''
      };
    }
    
    const latestVersion = release.tag_name;
    logger.info(`Latest version: ${latestVersion}`);
    
    // Compare versions
    const isNewer = isUpdateAvailable(currentVersion, latestVersion);
    
    if (isNewer) {
      logger.info('Update available!');
      return {
        available: true,
        currentVersion,
        latestVersion,
        releaseUrl: release.html_url,
        changelog: release.body || 'No changelog available.',
        publishedAt: release.published_at
      };
    } else {
      logger.info('No updates available');
      return {
        available: false,
        currentVersion,
        latestVersion: currentVersion,
        releaseUrl: '',
        changelog: '',
        publishedAt: ''
      };
    }
  } catch (error) {
    logger.error('Error checking for updates:', error);
    
    // Return no update available on error
    return {
      available: false,
      currentVersion: app.getVersion(),
      latestVersion: app.getVersion(),
      releaseUrl: '',
      changelog: '',
      publishedAt: ''
    };
  }
}

/**
 * Gets the GitHub releases page URL
 */
export function getReleasesPageUrl(): string {
  return `https://github.com/${GITHUB_OWNER}/${GITHUB_REPO}/releases`;
}
